Configure CI via Buildkite

mokagio opened the pull request:
Integrates Buildkite to run the UI tests against iPhone 12 and latest iPad Pro
We may add more Simulator targets in the future, e.g. if we discover fiddly methods that behave differently across Simulator.

To test, verify that CI run and passed.

See also https://github.com/Automattic/ScreenObject/pull/6.

mokagio committed:
Configure CI via Buildkite

mokagio commented on .buildkite/commands/run-tests.sh:
```diff
@@ -0,0 +1,8 @@
+xcodebuild clean test \
+  -project 'XCUITestHelpers.xcodeproj' \
+  -scheme 'XCUITestHelpers' \
+  -destination "platform=iOS Simulator,name=$SIMULATOR_NAME,OS=14.5"
```
See rationale from [this comment](https://github.com/Automattic/ScreenObject/pull/6/files#r695335144) on why this doesn't use Fastlane and has the OS value hardcoded.

AliSoftware replied:
I wonder if this note ☝️ is worth a `# comment` in the `.sh` file itself 🤔

AliSoftware reviewed (approved):
I'm ok with the rationale to not use `fastlane` here in order to avoid the whole Gemfile / bundler dance just for a single task.

BuildKite CI is green, so 👍

mokagio committed:
Add comment regarding hardcoded iOS version in tests command

Thanks for the suggestion @AliSoftware.

See
https://github.com/Automattic/XCUITestHelpers/pull/4#discussion_r695601359

AliSoftware commented on .buildkite/commands/run-tests.sh:
```diff
@@ -0,0 +1,21 @@
+#
+# - We are using raw `xcodebuild` without extra tooling such as Fastlane,
+#   `xcpretty`, or `xcbeautify`. This is an intentional compromise to keep the
+#   setup lean, as adding any of those would require extras in CI such as
+#   chating for what, right now, looks like merely a log readability gain.
```
Not sure I understand the wording of "require extras in CI such as chating"? I understand the original rationale (i.e it would require extra things like Gemfile and bundler setup) just not what "chating" means in this context, or is it a typo?

mokagio replied:
It's a typo. I meant "caching".

Still, I like your "tooling and ceremony" wording and I will adopt that. 👍

AliSoftware commented on .buildkite/commands/run-tests.sh:
```diff
@@ -0,0 +1,21 @@
+# - The iOS Simulator version is hardcoded in the command call, unlike the
+#   Simulator device, so we have a single source of truth. The downside is that
+#   the syntax in the pipeline is less clear: a reader might as where's the
+#   iOS version set. As long as we only need to test against the latest iOS
```
```suggestion
#   the syntax in the pipeline is less clear: a reader might ask where the
#   iOS version is set. As long as we only need to test against the latest iOS
```

AliSoftware replied:
I was having trouble understanding the rationale for that, as opposed to declare a `SIMULATOR_VERSION=14.5` constant on line 17 of the script and use that constant on line 21.
Then I realized that what you meant was probably "as opposed to declare `SIMULATOR_VERSION=$2` and allow callers of that script to provide a custom version.

💄 nit: I am ok with making the simulator version not be a parameter to keep a single source of truth. But I think it could be more readable to extract that as a `SIMULATOR_VERSION=14.5` constant at the top of the file for readability and ease of maintenance 😉

AliSoftware commented on .buildkite/commands/run-tests.sh:
```diff
@@ -0,0 +1,21 @@
+xcodebuild clean test \
+  -project 'XCUITestHelpers.xcodeproj' \
+  -scheme 'XCUITestHelpers' \
+  -destination "platform=iOS Simulator,name=$SIMULATOR_NAME,OS=14.5"
```
Full edit suggestion:
```suggestion
# Notice that:
#
# - We are using raw `xcodebuild` without extra tooling such as Fastlane,
#   `xcpretty`, or `xcbeautify`. This is an intentional compromise to keep the
#   setup lean, as adding any of those would require extra tooling and ceremony
#   in CI for what, right now, merely looks like a log readability gain.
#
# - The iOS Simulator version is hardcoded in the script (unlike the device),
#   so that we have a single source of truth. The downside is that the syntax
#   in the pipeline is less clear: a reader might wonder where the iOS version
#   is set. As long as we only need to test against the latest iOS version,
#   that seems like a reasonable tradeoff to make it easier to move to newer
#   versions as they are released.

SIMULATOR_NAME=$1
SIMULATOR_VERSION=14.5

xcodebuild clean test \
  -project 'XCUITestHelpers.xcodeproj' \
  -scheme 'XCUITestHelpers' \
  -destination "platform=iOS Simulator,name=$SIMULATOR_NAME,OS=$SIMULATOR_VERSION"
```

AliSoftware replied:
btw: disabled auto-merge on this PR to let you the chance to review this suggestion ☝️ first and decide if you want to apply it or not before merging (as I didn't think it warranted a "Request changes" review just for that and don't want to block this PR for a wording nitpick like this)

AliSoftware commented:
Also I noticed that the BuildKite jobs failed with the following error:
```
client_loop: send disconnect: Broken pipe
--
  | 🚨 Error: The command exited with status 255
  | user command error: The global command hook exited with status 255
```
If that's not just a seldom fluke we might cave in and switch to fastlane after all, which typically handles all those potential flukes that can happen with `xcodebuild`. Let's see if it's frequent enough to warrant that first tho, in case we can keep the raw `xcodebuild` still.

---

Additionally, if we keep the raw `xcodebuild`, we could consider using the `-quiet` flag on `xcodebuild` to reduce the CI log verbosity to a minimum? (tho maybe it would be _too_ minimalist?)

pachlava reviewed (approved):
Thank you for including me as a reviewer, so that I get to see the glimpses of Buildkite configuration. I have nothing to add to already existing review notes. 

As for the `Error: The command exited with status 255` - it's hard for me to judge if this is enough for considering a different approach (like using fastlane as @AliSoftware mentions).

I didn't want to rerun the steps since I wasn't sure if the original error will be kept in history.

mokagio commented:
Hey @AliSoftware @pachlava just wanted to leave a comment to say thanks for the review and the feedback.

I'll close the loop on this PR next week, as I'm currently on support rotation. 👌

mokagio committed:
Extract Buildkite simulator version in script constant + reword comment

Hat tip @AliSoftware.

See
https://github.com/Automattic/XCUITestHelpers/pull/4#discussion_r696399590

mokagio committed:
Add `-quiet` option to `xcodebuild` CI invocation to keep logs short

With this option, we won't get any build information and will get output
only when a test fails.

mokagio commented:
Thanks again for the review folks 🙇‍♂️ .

@AliSoftware I took onboard your suggestions, thank you. I agree, let's see if the `client_loop: send disconnect: Broken pipe` issue persist. If that's the case, I'll happily setup Fastlane because it's overhead will be more than justified.
